Add-Ons Webserver

Noordfrees opened the pull request:
The counterpart to https://github.com/widelands/wl_addons_server/pull/58 (now merged).

This allows the add-ons manager to connect to an add-ons web server. By default Widelands connects to the official server.

This implements a lot of goodies like commenting on add-ons, rating them, uploading add-ons, uploading screenshots, much better download speeds (no way to know yet much of this bonus is deducted when the server does not run on localhost…), and a whole bunch of other nice stuff. Oh and I got rid of CURL again, we're now using a native low-level backend instead (I hope Windows and MacOS like it, those are 100% untested).

Noordfrees committed:
Start to implement connecting to a webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Implement i18n, upgrading, votings, and comments

Noordfrees committed:
Use server for screenies as well

Noordfrees committed:
Remove CURL dependency

Noordfrees committed:
Adapt to new server protocol

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Refactor login handling and protocol

Noordfrees committed:
Merged master

Noordfrees committed:
Representative icons

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Fix multiline commenting

Noordfrees committed:
Merge ../widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
ProgressIndicatorWindow enhancements

Noordfrees committed:
Implement uploading add-ons

Noordfrees committed:
Merge ../widelands into addons-webserver

Noordfrees committed:
Merge ../widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Recover more gracefully after a protocol error

Noordfrees committed:
Better login box

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Confirm before upload

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Implement screenie upload

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Misc upload-related fixes

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge ../widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Update protocol

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Handle broken pipes elegantly

Noordfrees committed:
Check for invalid strings in headers

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Add label to screenie upload window

Noordfrees committed:
Fix connecting without login and getting votes when not logged in

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Update comment protocol

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Exclude line numbers from expected errors

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Add contact form

Noordfrees committed:
Enable contact form only for registered users, and offer only PM reply

Noordfrees committed:
Merge ../widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Add uploader to filter criteria

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Better error messages

Noordfrees committed:
Rephrase the helpful text wall

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Update doc

Noordfrees committed:
Allow editing comments

Noordfrees committed:
Nice graphical comment editor with previews

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Add comment editor RT markup buttons

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Guard against thread criss-crossing

Noordfrees committed:
Fix assert fail

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Make progress indicator pseudo-modal again

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Cleanup curl dependencies

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Doc enhancements

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge ../widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
6 files were automatically formatted.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Clang-tidy

Noordfrees committed:
'src/ui_fsmenu/addons.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Fix clang-tidy false-positive

Noordfrees committed:
Make mac and windows happy

Noordfrees committed:
'src/ui_fsmenu/addons.cc' was automatically formatted.

Noordfrees committed:
Next try for windows

Noordfrees committed:
'src/network/net_addons.cc' was automatically formatted.

Noordfrees committed:
Next try for windows

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
'src/ui_basic/panel.h' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into addons-webserver

Noordfrees committed:
Merge branch 'master' into addons-webserver

Noordfrees committed:
Merge branch 'master' into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
'src/ui_fsmenu/addons.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Fix bug with installing add-on translations

Noordfrees committed:
'src/ui_fsmenu/addons.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'master' into addons-webserver

Noordfrees committed:
2 files were automatically formatted.

Noordfrees committed:
Merge branch 'master' into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' into addons-webserver

Noordfrees committed:
'src/scripting/lua_map.cc' was automatically formatted.

Noordfrees committed:
Merge branch 'addons-webserver' of github.com:Noordfrees/widelands into addons-webserver

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

matthiakl commented:
In general everything works fine. The download speed is great and I like the Screenshot selection window :+1:.
My results:
- After submitting a comment only its headline is visible, the content only shows up after reopening the window
- After uploading an add-on, the metadata files are not updated. Therefore the old version is shown and it is still verified.
- Special characters in the contact form are not encoded properly (try üöä). Maybe add a hint to write the enquiry in English.
- After testing around I got 2 ASAN errors which might be related:
```
Direct leak of 528 byte(s) in 1 object(s) allocated from:
    #0 0x7fb28a987ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55c6b44532ab in AddOns::preload_addon(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /mnt/storage/Programme/widelands-dev/src/logic/addons.cc:403
    #2 0x55c6b3ef25d8 in WLApplication::initialize_g_addons() /mnt/storage/Programme/widelands-dev/src/wlapplication.cc:484
    #3 0x55c6b3ef0cce in WLApplication::WLApplication(int, char const* const*) /mnt/storage/Programme/widelands-dev/src/wlapplication.cc:407
    #4 0x55c6b3eef3ff in WLApplication::get(int, char const**) /mnt/storage/Programme/widelands-dev/src/wlapplication.cc:302
    #5 0x55c6b3eec99d in main /mnt/storage/Programme/widelands-dev/src/main.cc:45
    #6 0x7fb2881b5b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
    
Indirect leak of 16 byte(s) in 1 object(s) allocated from:
    #0 0x7fb28a987ca1 in operator new(unsigned long) /build/gcc/src/gcc/libsanitizer/asan/asan_new_delete.cpp:99
    #1 0x55c6b43b7ed1 in __gnu_cxx::new_allocator<unsigned int>::allocate(unsigned long, void const*) /usr/include/c++/11.1.0/ext/new_allocator.h:121
    #2 0x55c6b43b4fa5 in std::allocator_traits<std::allocator<unsigned int> >::allocate(std::allocator<unsigned int>&, unsigned long) /usr/include/c++/11.1.0/bits/alloc_traits.h:460
    #3 0x55c6b43b0539 in std::_Vector_base<unsigned int, std::allocator<unsigned int> >::_M_allocate(unsigned long) /usr/include/c++/11.1.0/bits/stl_vector.h:346
    #4 0x55c6b4463239 in void std::vector<unsigned int, std::allocator<unsigned int> >::_M_realloc_insert<unsigned int>(__gnu_cxx::__normal_iterator<unsigned int*, std::vector<unsigned int, std::allocator<unsigned int> > >, unsigned int&&) /usr/include/c++/11.1.0/bits/vector.tcc:440
    #5 0x55c6b445f1e2 in void std::vector<unsigned int, std::allocator<unsigned int> >::emplace_back<unsigned int>(unsigned int&&) /usr/include/c++/11.1.0/bits/vector.tcc:121
    #6 0x55c6b445be5f in std::vector<unsigned int, std::allocator<unsigned int> >::push_back(unsigned int&&) /usr/include/c++/11.1.0/bits/stl_vector.h:1204
    #7 0x55c6b444a5ff in AddOns::string_to_version(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) /mnt/storage/Programme/widelands-dev/src/logic/addons.cc:101
    #8 0x55c6b44534f7 in AddOns::preload_addon(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) /mnt/storage/Programme/widelands-dev/src/logic/addons.cc:408
    #9 0x55c6b3ef25d8 in WLApplication::initialize_g_addons() /mnt/storage/Programme/widelands-dev/src/wlapplication.cc:484
    #10 0x55c6b3ef0cce in WLApplication::WLApplication(int, char const* const*) /mnt/storage/Programme/widelands-dev/src/wlapplication.cc:407
    #11 0x55c6b3eef3ff in WLApplication::get(int, char const**) /mnt/storage/Programme/widelands-dev/src/wlapplication.cc:302
    #12 0x55c6b3eec99d in main /mnt/storage/Programme/widelands-dev/src/main.cc:45
    #13 0x7fb2881b5b24 in __libc_start_main (/usr/lib/libc.so.6+0x27b24)
```

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

matthiakl commented:
Windows does not seem to like it out of the box, I get a connection error:
Connection interrupted (Bad file descriptor).
The server is up and receives the request:
```
[Fri Jul 30 18:35:54 CEST 2021 @ Thread-1] Connection received from /192.168.0.148 : 56847 on /192.168.0.10:7399 (/192.168.0.10 : 7399).
```
It might be an issue with my setup, so I can try another one with multiple machines. Currently I have the SQL server running in a LXC container, the addon server on the host and Win10 inside VirtualBox :-)

Noordfrees committed:
Fix updating of comments window and metadata of new uploads

Noordfrees committed:
Fix memory leaks

Noordfrees commented:
All issues should be fixed now except the windows one. I'll try to have a look at that one with my windows machine tomorrow

matthiakl commented:
The comment preview does not split long lines automatically and if a line gets too long the game crashes:
```
[00:17:01.698 real] ERROR: Texture (2059, 18) too big! Maximum size is 2048.
terminate called after throwing an instance of 'RT::TextureTooBig'
  what():  Texture (2059, 18) too big! Maximum size is 2048.
```

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees commented:
I tried this and it seems that only ultra-long words without any whitespaces are affected. That's an error in the font renderer and not related to this branch – just add a very long word to the `elemental` file of any map and select it in the New Game dialog and you get the same crash.

matthiakl commented:
For Windows you have to use the winsock functions. With the attached diff everything works as it should (MSVC not mingw compiler).
I suggest to define a wlsend/wlreceive macro at the top to choose between read/write or send/recv.
[winsock.diff.txt](https://github.com/widelands/widelands/files/6911232/winsock.diff.txt)

Edit: for reference https://docs.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-recv

Noordfrees committed:
Fix windows-specific read/recv/write/send functions

Co-authored-by: matthiakl <t-m.42@arcor.de>

Noordfrees committed:
'src/network/net_addons.cc' was automatically formatted.

Noordfrees committed:
Add IP resolution and change default IP and port

Noordfrees committed:
Fix RemoteInteractionWindow layouting

Noordfrees committed:
Merge branch 'master' of github.com:widelands/widelands into addons-webserver

Noordfrees committed:
'src/network/net_addons.cc' was automatically formatted.

Noordfrees commented:
Fixed.
The actual server is up and running now, and I changed the default values to connect to the real server from now on. You should be able to access it and log in with your online gaming password. 
The IP address defaults to `widelands.org` (which is automatically resolved to `85.235.66.69`). IPv6 support seems to be trickier so I didn't add it for now; do we need this?
Btw I changed the recommended server port number to 7388 because 7399 is already used by the metaserver.
I also pushed another minor [codechange](https://github.com/widelands/wl_addons_server/commit/790371a6d066e2e0ef407123ee14ca5ede83ba1a) to the server to fix handling of the annoying double-encoded way the database stores the passwords.

matthiakl commented:
I verified the connection with Windows build from the Github action.
> The actual server is up and running now, and I changed the default values to connect to the real server from now on. You should be able to access it and log in with your online gaming password.

I get a permission error: 
```
[00:01:15.817 real] ERROR: set_login as 'matthiakl': access denied (Expected login result, received:
[00:01:15.817 real] ERROR: WL Protocol Exception: User matthiakl has no permissions at all) 
```
Maybe the permissions column is not populated?
Also the Metaserver does not accept the login anymore:
```
[00:06:51.091 real] INFO: InternetGaming: Sending login request.
[00:06:51.115 real] ERROR: [BufferedConnection] Error when receiving data from host (error 2: End of file)
[00:06:51.116 real] ERROR: [BufferedConnection] Closing socket
[00:06:51.116 real] INFO: [BufferedConnection] Stopping networking thread
[00:06:51.116 real] ERROR: InternetGaming: Error: Die Verbindung zum Metaserver wurde getrennt.
```
The credentials work fine on the website.

> The IP address defaults to widelands.org (which is automatically resolved to 85.235.66.69). IPv6 support seems to be trickier so I didn't add it for now; do we need this?

Should be fine for now.

Noordfrees commented:
According to the website admin site you did not set an online gaming password yet. Correct? For both the metaserver and the add-ons server you need to use your [online gaming password](https://www.widelands.org/ggz/changepw), not the website password.

matthiakl commented:
Silly me...
I thought it was the same. Now it works.

matthiakl reviewed (approved):
Let's have it.
I think the UI from `addons.cc` should be split into multiple files in the future.

Noordfrees commented:
Many thanks for your review of this large feature! :)
